plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerContentProvider.java

```diff
@@ -56,16 +57,37 @@ protected void handleConnectionChanged(IConnection connection, String property,
 				refreshGrouping(groupMap.get(resource.getProject()), resource.getKind());
 			}else if(oldValue != null && newValue == null) {
 				//delete
-				IResource resource = (IResource)oldValue;
+				IResource resource = (IResource) oldValue;
 				refreshGrouping(groupMap.get(resource.getProject()), resource.getKind());
 			}else {
 				refreshViewer(newValue);
 			}
+		}else if(ConnectionProperties.PROPERTY_PROJECTS.equals(property)){
+			handleProjectChanges(oldValue, newValue);
 		}else{
 			super.handleConnectionChanged(connection, property, oldValue, newValue);
 		}
 	}
 	
+	@SuppressWarnings("unchecked")
+	private void handleProjectChanges(Object oldValue, Object newValue) {
+		List<IProject> newProjects = (List<IProject>) newValue;
+		List<IProject> oldProjects = (List<IProject>) oldValue;
+		//theory this should only be one.
+		Object [] diffs = CollectionUtils.disjunction(newProjects, oldProjects).toArray();
+		if(newProjects.size() > oldProjects.size()) {
+			Connection connection = ConnectionsRegistryUtil.getConnectionFor(newProjects.get(0));
+			addChildrenToViewer(connection, diffs);
+		}else if(newProjects.size() < oldProjects.size()){
+			Connection connection = ConnectionsRegistryUtil.getConnectionFor(oldProjects.get(0));
+			removeChildrentFromViewer(connection, diffs);
+			//is this safe across diffs?
+			for (Object project : diffs) {
+				groupMap.remove(project);
+			}
+		}
```

Comment:
As far as I know in databinding if an element of the list is changed while the list by itself wont (no elements removed/added) then no listChange event happens. Afaik a change in an element is - in databinding - catched by the element listener, not the list change listener. So depending on your change mgmt strategy Diff wont inform you sufficiently if you catch element changes in the list change event?